File: problem/Min_Cost_String_I/minimumCost.py
class Solution:
    def minimumCost(self, source, target, original, changed, cost):
        inf = float('inf')
        dist = [[inf] * 26 for _ in range(26)]
        
        for i in range(26):
            dist[i][i] = 0
            
        for o, c, w in zip(original, changed, cost):
            u = ord(o) - ord('a')
            v = ord(c) - ord('a')
            dist[u][v] = min(dist[u][v], w)
            
        for k in range(26):
            for i in range(26):
                if dist[i][k] == inf:
                    continue
                for j in range(26):
                    if dist[k][j] != inf:
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
                        
        total_cost = 0
        for s_char, t_char in zip(source, target):
            u = ord(s_char) - ord('a')
            v = ord(t_char) - ord('a')
            if u == v:
                continue
            if dist[u][v] == inf:
                return -1
            total_cost += dist[u][v]
            
        return total_cost

File: problem/Min_Cost_String_I/minimumCost.c
#include <string.h>
#include <limits.h>

long long minimumCost(char *source, char *target, char *original, int originalSize, char *changed, int changedSize, int *cost, int costSize)
{
    long long dist[26][26];
    const long long INF = 1e14;

    for (int i = 0; i < 26; ++i)
    {
        for (int j = 0; j < 26; ++j)
        {
            dist[i][j] = (i == j) ? 0 : INF;
        }
    }

    for (int i = 0; i < originalSize; ++i)
    {
        int u = original[i] - 'a';
        int v = changed[i] - 'a';
        if ((long long)cost[i] < dist[u][v])
        {
            dist[u][v] = (long long)cost[i];
        }
    }

    for (int k = 0; k < 26; ++k)
    {
        for (int i = 0; i < 26; ++i)
        {
            if (dist[i][k] == INF)
                continue;
            for (int j = 0; j < 26; ++j)
            {
                if (dist[k][j] != INF)
                {
                    if (dist[i][k] + dist[k][j] < dist[i][j])
                    {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }
    }

    long long totalCost = 0;
    int n = strlen(source);

    for (int i = 0; i < n; ++i)
    {
        int u = source[i] - 'a';
        int v = target[i] - 'a';
        if (u == v)
            continue;
        if (dist[u][v] == INF)
            return -1;
        totalCost += dist[u][v];
    }

    return totalCost;
}